Source/Diagnostics/Diagnostics.cpp

```diff
@@ -103,6 +103,57 @@ Diagnostics::BaseReadParameters ()
 
     bool species_specified = pp.queryarr("species", m_species_names);
 
+    // Prepare to dump rho per species
+    amrex::Vector<std::string> variables;
+    const MultiParticleContainer& mpc = warpx.GetPartContainer();
+    // If diag.species is not provided, loop over all species to check whether
+    // rho per species is requested
+    if (m_species_names.size() == 0) m_species_names = mpc.GetSpeciesNames();
+    // ns: total number of species
+    // ns_dump_rho: number of species that dump rho per species
+    const int ns = int(m_species_names.size());
+    int ns_dump_rho = 0;
+    // is: species index to loop over all species
+    // is_dump_rho: species index to loop over species that dump rho per species
+    int is;
+    int is_dump_rho;
+    // Loop over all species
+    for (is = 0; is < ns; is++) {
```

Comment:
```suggestion
    for (int is = 0; is < ns; is++) {
```

Reply:
You can also make this a nice range-based loop:
```suggestion
    for (const auto & species_name : m_species_names) {
```